dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/BitBucketSourceCodeRepo.java

```diff
@@ -212,24 +247,55 @@ public void updateReferenceType(String repositoryId, Version version) {
         RefsApi refsApi = new RefsApi(apiClient);
         // There isn't exactly a single Bitbucket endpoint to get a version which then allows us to determine if it's a branch or tag.
         // This code checks two endpoints (branches and tags) to see if it belongs in which.
-        try {
-            refsApi.repositoriesUsernameRepoSlugRefsBranchesNameGet(workspace, name, repoSlug);
-            version.setReferenceType(Version.ReferenceType.BRANCH);
-            return;
-        } catch (ApiException e) {
-            LOG.warn(gitUsername + ": apiexception on reading branches" + e.getMessage(), e);
-            // this is not so critical to warrant a http error code
+        boolean rateLimited;
+        do {
+            try {
+                refsApi.repositoriesUsernameRepoSlugRefsBranchesNameGet(workspace, name, repoSlug);
+                version.setReferenceType(Version.ReferenceType.BRANCH);
+                return;
+            } catch (ApiException e) {
+                rateLimited = isRateLimited(false, e, "reading branches");
+            }
+
+            try {
+                refsApi.repositoriesUsernameRepoSlugRefsTagsNameGet(workspace, name, repoSlug);
+                version.setReferenceType(Version.ReferenceType.TAG);
+                return;
+            } catch (ApiException e) {
+                // this is not so critical to warrant a http error code
+                rateLimited = isRateLimited(rateLimited, e, "reading tags");
+            }
+        } while (rateLimited);
+
+        throw new CustomWebApplicationException(name + " is not a Bitbucket branch or tag in " + repositoryId, HttpStatus.SC_INTERNAL_SERVER_ERROR);
+    }
+
+    /**
+     * determine whether a request is rate-limited, output logging mesages, and potentially sleep past rate-limiting
+     * @param rateLimited
+     * @param e
+     * @param message
+     * @return
+     */
+    private boolean isRateLimited(final boolean rateLimited, ApiException e, String message) {
+        if (LOG.isWarnEnabled()) {
+            LOG.warn("%s: apiexception on %s%s".formatted(gitUsername, message, e.getMessage()), e);
         }
+        // this is not so critical to warrant a http error code
+        boolean newlyRateLimited = false;
+        if (e.getCode() == Status.TOO_MANY_REQUESTS.getStatusCode()) {
+            newlyRateLimited = true;
+            LOG.error("%s: rate-limited on %s".formatted(gitUsername, message));
+            try {
 
-        try {
-            refsApi.repositoriesUsernameRepoSlugRefsTagsNameGet(workspace, name, repoSlug);
-            version.setReferenceType(Version.ReferenceType.TAG);
-            return;
-        } catch (ApiException e) {
-            LOG.warn(gitUsername + ": apiexception on reading tags" + e.getMessage(), e);
-            // this is not so critical to warrant a http error code
+                Thread.sleep(ONE_MINUTE_IN_MS); // one minute, should be exponential back-off
```

Comment:
I'm a little paranoid about this; if we get a flood of Bitbucket requests it could end up blocking a lot of threads. But in practice it's probably OK.

Reply:
Going to deal with this using the queue approach, recommend reverting the change in this file specifically in 1.14 
https://ucsc-cgl.atlassian.net/browse/SEAB-5192